storage/wal/dump.go

```diff
@@ -0,0 +1,194 @@
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"text/tabwriter"
+)
+
+// Command represents the program execution for "influxd inspect dumptsmwal
+// This command will dump all entries from a given list of wal files
+type Dump struct {
+	// Standard input/output, overridden for testing.
+	Stderr io.Writer
+	Stdout io.Writer
+
+	// A list of files to dump
+	Files []string
+
+	// Whether or not to check for duplicate/out of order entries
+	FindDuplicates bool
+}
+
+type DumpReport struct {
+	File          string
+	DuplicateKeys []string
+}
+
+// Run executes the command.
+func (w *Dump) Run(print bool) ([]*DumpReport, error) {
+	if w.Stderr == nil {
+		w.Stderr = os.Stderr
+	}
+
+	if w.Stdout == nil {
+		w.Stdout = os.Stdout
+	}
+
+	if !print {
+		w.Stdout, w.Stderr = ioutil.Discard, ioutil.Discard
+	}
+
+	tw := tabwriter.NewWriter(w.Stdout, 8, 2, 1, ' ', 0)
+
+	// Process each TSM WAL file.
+
+	var reports []*DumpReport
+	for _, path := range w.Files {
+		duplicateKeys, err := w.process(tw, path)
+		if err != nil {
+			return nil, err
+		}
+
+		r := &DumpReport{
+			File:          path,
+			DuplicateKeys: duplicateKeys,
+		}
+		reports = append(reports, r)
+	}
+
+	return reports, nil
+}
+
+func (w *Dump) process(out io.Writer, path string) ([]string, error) {
+	if filepath.Ext(path) != "."+WALFileExtension {
+		fmt.Fprintf(out, "invalid wal filename, skipping %s", path)
+		return nil, nil
+	}
+
+	fmt.Fprintf(out, "File: %s\n", path)
+
+	// Track the earliest timestamp for each key and a set of keys with out-of-order points.
+	minTimestampByKey := make(map[string]int64)
+	duplicateKeys := make(map[string]struct{})
+
+	// Open WAL reader.
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	r := NewWALSegmentReader(f)
+
+	// Iterate over the WAL entries.
+	for r.Next() {
+		entry, err := r.Read()
+		if err != nil {
+			return nil, fmt.Errorf("cannot read entry: %s", err)
+		}
+
+		switch entry := entry.(type) {
+		case *WriteWALEntry:
+			if !w.FindDuplicates {
+				fmt.Fprintf(out, "[write] sz=%d\n", entry.MarshalSize())
+			}
+
+			keys := make([]string, 0, len(entry.Values))
+			for k := range entry.Values {
+				keys = append(keys, k)
+			}
+			sort.Strings(keys)
+
+			for _, k := range keys {
+				fmtKey, err := formatKeyOrgBucket(k)
+				// if key cannot be properly formatted with org and bucket, skip printing
+				if err != nil {
+					fmt.Fprintf(out, "Error: %v\n", err)
+					continue
+				}
+				for _, v := range entry.Values[k] {
+					t := v.UnixNano()
+
+					// Check for duplicate/out of order keys.
+					if min, ok := minTimestampByKey[k]; ok && t <= min {
+						duplicateKeys[k] = struct{}{}
+					}
+					minTimestampByKey[k] = t
+
+					// Skip printing if we are only showing duplicate keys.
+					if w.FindDuplicates {
+						continue
+					}
+
+					switch v := v.(type) {
+					case value.IntegerValue:
+						fmt.Fprintf(out, "%s %vi %d\n", fmtKey, v.Value(), t)
```

Comment:
These will need to be added to the report summary, so that they can be examined programmatically.